File: beyondnodes.m
function [beyond_node_degree,asymm_links] = beyondnodes(inputs,sort_idx,max_match_num,asymm_links)
 %%找出并标记每条链接上的节点及节点超出节点度的情况
   beyond_nodes = [];
   for node_i = 1:inputs.nodes_num
      
        [node_i_rows_asy_1{node_i},~] = find(asymm_links(:,1) == sort_idx(node_i)); 
        [node_i_rows_asy_2{node_i},~] = find(asymm_links(:,2) == sort_idx(node_i)); 
        node_i_rows_asy{node_i} = [node_i_rows_asy_1{node_i};node_i_rows_asy_2{node_i}];
        beyond_node_degree(node_i) = length(node_i_rows_asy{node_i}) - max_match_num(sort_idx(node_i));
    
        %%node_i实际上是排序后的索引
        %%根据链接的两个节点是否节点度都超出对链接进行排序
        if beyond_node_degree(node_i) > 0
           beyond_nodes = [beyond_nodes,sort_idx(node_i)];
        end
   end 
    %%对链接进行重新排序
    num_b = sum(ismember(asymm_links, beyond_nodes), 2);
    [~, sorted_Indices] = sort(num_b, 'descend');
    asymm_links = asymm_links(sorted_Indices, :);
end

File: Beyond_Nodes.py
# -*- coding:utf-8 -*-
"""
作者：TBJ
日期：2024年06月27日
"""
import numpy as np


def beyond_nodes(inputs, sort_idx, max_match_num, asy_mm_links):
    beyondNodes = []
    node_i_rows_asy = np.empty(inputs.nodes_num, dtype=object)
    beyond_node_degree = np.empty(inputs.nodes_num, dtype=object)
    for node_i in range(0, inputs.nodes_num):
        node_i_rows_asy_1 = np.where([asy_mm_links[k][0] for k in range(0, len(asy_mm_links))] == sort_idx[node_i])
        node_i_rows_asy_1 = node_i_rows_asy_1[0]
        node_i_rows_asy_2 = np.where([asy_mm_links[k][1] for k in range(0, len(asy_mm_links))] == sort_idx[node_i])
        node_i_rows_asy_2 = node_i_rows_asy_2[0]
        node_i_rows_asy[node_i] = np.concatenate((node_i_rows_asy_1, node_i_rows_asy_2))
        beyond_node_degree[node_i] = len(node_i_rows_asy[node_i]) - max_match_num[sort_idx[node_i]]

        if beyond_node_degree[node_i] > 0:
            beyondNodes.append(sort_idx[node_i])

    num_b = np.zeros(len(asy_mm_links))

    for i in range(0, len(asy_mm_links)):
        for k in range(0, len(asy_mm_links[i])):
            num_b += asy_mm_links[i][k] in beyondNodes

    sorted_Indices = np.argsort(- num_b)
    asy_mm_links = [asy_mm_links[sorted_Indices[i]] for i in range(0, len(sorted_Indices))]

    return beyond_node_degree, asy_mm_links
